### PR TITLE
feat(zql): Implement OR

### PR DESCRIPTION
**Pull request by arv**
This supersedes #52

OR is implemented by branching the graph when an OR expression is encountered. Each OR branch is executed "in parallel" and then merged together. The merging is done using `concat` and then `distinct`. 

For example if we have:

```sql
SELECT * FROM table WHERE (a=1 OR b=1) AND (a=2 OR b=2)
```

We get something like:

```mermaid
  graph TD;
      Root-->id1;
      Root-->id2;
      id1(a=1) --> c1;
      id2(b=1) --> c1;
      c1(Concat) --> d1;
      d1(Distinct) --> id3;
      d1 --> id4;
      id3(a=2) --> c2;
      id4(b=2) --> c2;
      c2(Concat) --> d2;
      d2(Distinct)


```

### CODE REVIEW

**Comment by arv** [comment]
I have builder changes but that can be reviewed in isolation.

File: src/zql/ivm/graph/operators/distinct-operator.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    if (!emittedMap) {
+      emittedMap = new Map();
+      this.#emitted.set(version, emittedMap);
+    }
+
+    const newMultiset: Entry<T>[] = [];
+    for (const [value, multiplicity] of entriesForThisVersion.values()) {
+      const {id} = value;
+      const existingMultiplicity = emittedMap.get(id) ?? 0;
+      const desiredMultiplicity = Math.sign(multiplicity);
```
Comment (aboodman) [inline src/zql/ivm/graph/operators/distinct-operator.ts:57]: Totally forgot about `Math.sign`.

File: src/zql/ivm/graph/operators/distinct-operator.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  // number is the multiplicity outputted so far in this tx.
+  readonly #emitted = new Map<Version, Map<string, number>>();
+
+  constructor(input: DifferenceStream<T>, output: DifferenceStream<T>) {
+    super(input, output, (version, data) => this.#handleDiff(version, data));
+  }
+
+  #handleDiff(version: number, multiset: Multiset<T>): Multiset<T> {
+    // Clear data for old versions.
+    clearOldVersions(this.#entriesCache, version);
```
Comment (aboodman) [inline src/zql/ivm/graph/operators/distinct-operator.ts:27]: I thought that each 'version' would call `handleDiff` one time only. If so, couldn't the cache just be a local?

Comment (arv) [reply src/zql/ivm/graph/operators/distinct-operator.ts:27]: Maybe. I'm trying to understand if that is the case.

@tantaman ?

**Comment by tantaman** [comment]
img

File: src/zql/ivm/graph/operators/concat-operator.test.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  inputs[1].commit(version);
+  inputs[2].commit(version);
+  expect(items).toEqual([[[{x: 0}, 1]], [[{x: 1}, 1]], [[{x: 2}, 2]]]);
+});
+
+test('Test with single input', () => {
+  type T = {x: number};
+  const input = new DifferenceStream<T>();
+
+  const output = input.concat();
```
Comment (tantaman) [inline src/zql/ivm/graph/operators/concat-operator.test.ts:51]: does it make sense to concat with nothing?

File: src/zql/ivm/graph/operators/distinct-operator.ts (hunk 2 above)
Comment (tantaman) [reply src/zql/ivm/graph/operators/distinct-operator.ts:27]: > I thought that each 'version' would call handleDiff one time only.

Not now that we optimistically send data down as soon as we get it (no waiting at operators for inputs on both edges).

image

add will notify depth first down the left then the right.

File: src/zql/ivm/graph/operators/distinct-operator.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        newMultiset.push([value, desiredMultiplicity - existingMultiplicity]);
+        emittedMap.set(id, desiredMultiplicity);
+      }
+    }
+
+    return newMultiset;
+  }
+}
+
+function clearOldVersions(m: Map<Version, unknown>, version: Version) {
```
Comment (tantaman) [inline src/zql/ivm/graph/operators/distinct-operator.ts:68]: This map should only ever be a single entry of the prior version. Can you confirm and only keep a single map and clearing it whenever `version > lastVersionSeen`?

e.g.,
```ts
#lastSeenVersion: number;

#handleDiff() {
  if (version > this.#lastSeenVersion) {
    // clear cache
  }
}
```

Comment (arv) [reply src/zql/ivm/graph/operators/distinct-operator.ts:68]: Good point

File: src/zql/ivm/graph/operators/concat-operator.test.ts (hunk 3 above)
Comment (arv) [reply src/zql/ivm/graph/operators/concat-operator.test.ts:51]: Not really. We can enforce the length using types.

I might remove this method and provide a function instead. As you can see, I split the array and then join it later which useless work with the sole purpose of fitting this API.

Comment (arv) [reply src/zql/ivm/graph/operators/concat-operator.test.ts:51]: Removed method in favor of function.

I didn't add any code/or types to ensure the length. `concat([])` and `concat([s])` works as expected even if not very useful.

File: src/zql/integration.test.ts [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
 import {expect, test} from 'vitest';
 import {z} from 'zod';
 import {generate} from '../generate.js';
 import {makeReplicacheContext} from './context/replicache-context.js';
-import {Replicache, TEST_LICENSE_KEY} from 'replicache';
-import {nanoid} from 'nanoid';
-import fc from 'fast-check';
-import {EntityQuery} from './query/entity-query.js';
 import * as agg from './query/agg.js';
+import {EntityQuery, expression, or} from './query/entity-query.js';
```
Comment (aboodman) [inline src/zql/integration.test.ts:9]: wooo.

Nit: I think your earlier idea of `exp` was a better starting point.

**Pull request merged**